### PR TITLE
integrates witness proof into public facade

### PR DESCRIPTION
**Pull request by kjezek**
This PR extends the public API of the  `WitnessProof` support. 

* This defines a new package `witness` with a simple `Proof` interface. 
* This interface is implemented by `mpt.WitnessProof`, i.e. the MPT support of the witness proof. 
* The `carmen.HistoricBlockContext` is extended of a new  features to extract the proof, i.e. a new method `GetProof`
* `State` exposes a method  `CreateWitnessProof` called from `carmen.HistoricBlockContext` and it provides `witness.Proof`, all schemes except S5 do `panic` when calling this method, only the `ArchiveTrie` supports it.  

An example is provided showing how to use.

### CODE REVIEW

File: go/carmen/carmen.go [hunk 1]
```diff
@@ -206,6 +217,7 @@ type HistoricBlockContext interface {
 // context before another one can be opened.
 // Methods of this interface are not thread safe.
 type TransactionContext interface {
+	QueryContext
```
Comment (petr-hanzl) [inline go/carmen/carmen.go:220]: I believe `TransactionContext` implicitly implements `QueryContext` as all method within `QrCtx` are contained within `TxCtx` hence this is not necessary

File: go/carmen/example_test.go [hunk 2]
```diff
@@ -282,3 +282,134 @@ func ExampleDatabase_GetHistoricContext() {
...
+
+	// serialise to a format to be passed to others
+	stream := carmen.MptWitnessProofToNodes(completeProof)
+	// recover at the other side
+	recoveredProof := carmen.MptWitnessProofFromNodes(stream)
+
+	// ------- Values can be proven outside this client  -------
+
+	// Prove the values for each address and key offline without access to database
+	rootHashes := make([]carmen.Hash, N) // TODO get real root hashes
```
Comment (petr-hanzl) [inline go/carmen/example_test.go:348]: nit: todo?

File: go/carmen/example_test.go [hunk 3]
```diff
@@ -282,3 +282,134 @@ func ExampleDatabase_GetHistoricContext() {
...
+
+	if err := db.Close(); err != nil {
+		log.Fatalf("cannot close db: %v", err)
+	}
+
+	if err := os.RemoveAll(dir); err != nil {
+		log.Fatalf("cannot remove dir: %v", err)
+	}
+
+	// Output: TODO
```
Comment (petr-hanzl) [inline go/carmen/example_test.go:414]: nit: todo?

File: go/database/mpt/proof.go [hunk 4]
```diff
@@ -297,3 +298,121 @@ func createPathFromKeyPrefix(key common.Key, nibbles uint8) Path {
...
+
+// AllAddressesEmpty checks that all accounts are empty for the input root hash and the address range.
+// If the witness proof contains all empty accounts for the input address range, it returns true.
+// An empty account is an account that contains a zero balance, nonce, and code hash.
+// If there is at least one non-empty account, it returns false. If the proof is not complete,
+// it returns unknown. An incomplete proof is a proof where the input address terminates in a node
+// that is not a correct account node.
+func (c *mptWitnessProofState) AllAddressesEmpty(from, to common.Address) (tribool.Tribool, error) {
+	return c.proof.AllAddressesEmpty(c.root, from, to)
+}
```
Comment (petr-hanzl) [inline go/database/mpt/proof.go:418]: I dont think the documentation here is needed. It seems like a copy of the documentation of the implemented interface `witness.Proof`.

**Review by petr-hanzl — COMMENTED** [review]
Looks good. Consider removing the documentation of `mptWitnessProofState`.

File: go/carmen/carmen.go [hunk 5]
```diff
@@ -188,6 +189,16 @@ type HeadBlockContext interface {
 type HistoricBlockContext interface {
 	blockContext
 
+	// CreateWitnessProof creates a witness proof for the given account and keys.
+	// Error may be produced when it occurs in the underlying database,
+	// otherwise, the proof is returned.
+	CreateWitnessProof(address Address, keys ...Key) (WitnessProof, error)
```
Comment (HerbertJordan) [inline go/carmen/carmen.go:195]: How about renaming this to just `GetProof`?

File: go/common/witness/proof.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	IsValid() bool
+
+	// Add adds a proof node to the witness proof.
+	Add(Proof)
+}
+
+// ProofState allows for getting properties of accounts and account storage slots.
+// It contains methods to determine values of properties based on addresses and keys,
+// and returns the values and information if the proof was sufficient to determine the value.
+type ProofState interface {
```
Comment (HerbertJordan) [inline go/common/witness/proof.go:22]: Why the differentiation between a `Proof` and a `ProofState`. Could this not be the same?

File: go/database/mpt/proof.go [hunk 7]
```diff
@@ -297,3 +298,121 @@ func createPathFromKeyPrefix(key common.Key, nibbles uint8) Path {
...
 }
+
+// NewWitnessProof creates a new witness proof.
+func NewWitnessProof() witness.Proof {
+	return witnessProofAdapter{WitnessProof{proofDb{}}}
+}
+
+// WitnessProofToNodes converts the witness proof to a list of strings.
+// Each string is an RLP node of the witness proof.
+func WitnessProofToNodes(proof witness.Proof) []string {
```
Comment (HerbertJordan) [inline go/database/mpt/proof.go:309]: I think this should be a member function of the proof instead of a stand-alone function.
It might also be part of the proof interface.

File: go/database/mpt/proof.go [hunk 8]
```diff
@@ -297,3 +298,121 @@ func createPathFromKeyPrefix(key common.Key, nibbles uint8) Path {
...
+	p := proof.(witnessProofAdapter).WitnessProof
```
Comment (HerbertJordan) [inline go/database/mpt/proof.go:310]: The way it is now, it really only works for this one type of proof implementation and crash for all others.

**Review by HerbertJordan — COMMENTED** [review]
Before reviewing it in depth, I would like to discuss the idea behind the differentiation between a `Proof` and a `ProofState`. What is the intention behind this? Wouldn't it be simpler to merge those two interfaces into a single `Proof` interface?

File: go/common/witness/proof.go (hunk 6 above)
Comment (kjezek) [reply go/common/witness/proof.go:22]: I was at first sticked to the idea that you create a proof first --> you get `Proof`, and "later" you query the proof (for a root hash), which gives you the `ProofState` -> you query this. 

I think it sounds plausible to merge these two together, perhaps I will try to implement and see how would this look like

File: go/database/mpt/proof.go (hunk 7 above)
Comment (kjezek) [reply go/database/mpt/proof.go:309]: it would loos genericity, the idea was that `witness.Proof` serves for all schemes while the nodes are only in MPT.

I was also thinking about something more general, like a method `Serialise/Deserialise`, which would work with `[]byte` etc, it could then go to the interface.

File: go/database/mpt/proof.go (hunk 8 above)
Comment (kjezek) [reply go/database/mpt/proof.go:310]: right, it was meant to work this way.  If we can create a proof from other scheme, it will produce a `witness.Proof `as well, but if a user uses it to convert as an MPT witness proof, it crashes - it is considered a user mistake. He is suppose to use only a proof instance created from respective scheme.

File: go/carmen/carmen.go [hunk 9]
```diff
@@ -206,6 +212,7 @@ type HistoricBlockContext interface {
 // context before another one can be opened.
 // Methods of this interface are not thread safe.
 type TransactionContext interface {
+	QueryContext
```
Comment (HerbertJordan) [inline go/carmen/carmen.go:215]: This looks unrelated ... if so, please remove this line.

File: go/carmen/block.go [hunk 10]
```diff
@@ -137,6 +139,27 @@ func (c *archiveBlockContext) RunTransaction(run func(TransactionContext) error)
 	return runTransaction(c, run)
 }
 
+func (c *archiveBlockContext) GetProof(address Address, keys ...Key) (WitnessProof, error) {
+	c.lock.Lock()
+	if c.db == nil {
+		c.lock.Unlock()
+		return WitnessProof{}, fmt.Errorf("cannot get proof in invalid block context")
+	}
+	c.lock.Unlock()
```
Comment (HerbertJordan) [inline go/carmen/block.go:148]: I think this lock should be held until the end since otherwise the database underlying the `archiveState` may get closed while running this function.

File: go/carmen/proof.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+// CreateWitnessProofFromNodes creates a witness proof from a list of strings.
+// Each string is an RLP node of the witness proof.
+func CreateWitnessProofFromNodes(elements ...string) WitnessProof {
+	proof := mpt.CreateWitnessProofFromNodes(elements)
+	return WitnessProof{proof}
+}
+
+type WitnessProof struct {
```
Comment (HerbertJordan) [inline go/carmen/proof.go:16]: To enable future refactoring hidden from external API users, we should hide this behind an interface.

File: go/database/mpt/proof.go [hunk 12]
```diff
@@ -201,6 +213,14 @@ func (p WitnessProof) String() string {
 	return b.String()
 }
 
+func (p WitnessProof) GetElements() []string {
+	res := make([]string, 0, len(p.proofDb))
+	for _, v := range p.proofDb {
```
Comment (HerbertJordan) [inline go/database/mpt/proof.go:218]: nit: for a future update, we might consider storing the elements inside the proof as strings instead of []byte.

File: go/database/mpt/proof.go [hunk 13]
```diff
@@ -201,6 +213,14 @@ func (p WitnessProof) String() string {
 	return b.String()
 }
 
+func (p WitnessProof) GetElements() []string {
```
Comment (HerbertJordan) [inline go/database/mpt/proof.go:216]: nit: document

File: go/database/mpt/proof.go (hunk 12 above)
Comment (kjezek) [reply go/database/mpt/proof.go:273]: what is the benefit of using strings instead?

File: go/database/mpt/proof.go (hunk 4 above)
Comment (kjezek) [reply go/database/mpt/proof.go]: done differently at the end

File: go/carmen/carmen.go (hunk 5 above)
Comment (kjezek) [reply go/carmen/carmen.go]: done

File: go/carmen/carmen.go (hunk 9 above)
Comment (kjezek) [reply go/carmen/carmen.go]: ok

File: go/carmen/block.go (hunk 10 above)
Comment (kjezek) [reply go/carmen/block.go:148]: I guess we should prevent closing of database when a witness proof creation is running, similarly as we do for the head block state, because locking the whole method may be too coarse?

Comment (kjezek) [reply go/carmen/block.go:148]: Actually it seems it is already in-place:
https://github.com/Fantom-foundation/Carmen/blob/main/go/carmen/database.go#L207
and
https://github.com/Fantom-foundation/Carmen/blob/main/go/carmen/database.go#L264

File: go/carmen/proof.go (hunk 11 above)
Comment (kjezek) [reply go/carmen/proof.go]: ok, done

File: go/database/mpt/proof.go (hunk 13 above)
Comment (kjezek) [reply go/database/mpt/proof.go:271]: done

File: go/carmen/example_test.go (hunk 2 above)
Comment (kjezek) [reply go/carmen/example_test.go]: finished, fixed

File: go/carmen/example_test.go (hunk 3 above)
Comment (kjezek) [reply go/carmen/example_test.go]: finished, fixed

File: go/state/cppstate/cpp_state.go [hunk 14]
```diff
@@ -279,6 +280,10 @@ func (cs *CppState) Check() error {
 	return nil
 }
 
+func (cs *CppState) CreateWitnessProof(address common.Address, keys ...common.Key) (witness.Proof, error) {
+	panic("not implemented")
+}
```
Comment (petr-hanzl) [inline go/state/cppstate/cpp_state.go:285]: I think this 'not implemented' issues might get piled up - we had the same thing with `HasEmptyStorage` func and I believe we might be extending these interfaces in near future. 

How about we create a general error `ErrNotImplemented` and return it instead of panicking and we will have this prepared for future unimplemented functions within unused schemas?

File: go/state/gostate/go_state.go [hunk 15]
```diff
@@ -424,3 +425,7 @@ func (s *GoState) GetSnapshotVerifier(metadata []byte) (backend.SnapshotVerifier
 	}
 	return backend.NewComposedSnapshotVerifier(verifiers, partCounts), nil
 }
+
+func (s *GoState) CreateWitnessProof(address common.Address, keys ...common.Key) (witness.Proof, error) {
+	panic("not implemented")
```
Comment (petr-hanzl) [inline go/state/gostate/go_state.go:430]: Same case as above...

**Review by petr-hanzl — DISMISSED** [review]

File: go/state/cppstate/cpp_state.go (hunk 14 above)
Comment (kjezek) [reply go/state/cppstate/cpp_state.go:285]: We could do this, but I wonder what would be the difference. `panic` is for sure more severe, but these methods should not be ever called. When only the public API and supported configuration is called, there is no way to execute these methods. If the client bypasses this and calls something which is not supported, I do not see a lot of difference in which way he is informed about calling an unsupported feature.

Comment (petr-hanzl) [reply go/state/cppstate/cpp_state.go:285]: ic. Yeah, you are right. There is actually no way this gets called by the public API. Lets keep the panic.

File: go/database/mpt/proof.go (hunk 12 above)
Comment (HerbertJordan) [reply go/database/mpt/proof.go:274]: You can skip the conversion step here and there, which is always a memory allocation and copy.

File: go/carmen/block.go (hunk 10 above)
Comment (HerbertJordan) [reply go/carmen/block.go:148]: Alright, this may allow multiple proof-queries in parallel.

File: go/carmen/database_test.go [hunk 16]
```diff
@@ -845,6 +845,374 @@ func TestDatabase_CloseDB_Unfinished_Queries(t *testing.T) {
...
+				}
+				if got, want := value, (Value{byte(i), byte(j), byte(k)}); got != want {
+					t.Errorf("unexpected value, wanted %v, got %v", want, got)
+				}
+			}
+		}
+	}
+}
+
+func TestDatabase_CloseDB_Unfinished_ProofE_CannotCloseDb(t *testing.T) {
```
Comment (HerbertJordan) [inline go/carmen/database_test.go:1103]: nit:
```suggestion
func TestDatabase_CloseDB_Unfinished_Proof_CannotCloseDb(t *testing.T) {
```

File: go/carmen/database_test.go [hunk 17]
```diff
@@ -845,6 +845,374 @@ func TestDatabase_CloseDB_Unfinished_Queries(t *testing.T) {
...
+	go func() {
+		// attempts to close db must keep failing
+		if err := db.Close(); !errors.Is(err, errBlockContextRunning) {
+			t.Errorf("closing db should fail: %v", err)
+		}
+		done.Store(true)
+	}()
+
+	for !done.Load() {
+		_, err = block.GetProof(Address{1})
```
Comment (HerbertJordan) [inline go/carmen/database_test.go:1142]: This loop may never be executed if the go routine above is fast.

I don't think this test benefits from the concurrent execution of the close attempt and the reading of the proof. What would get lost if those steps would be done in sequence:
- attempt to close the db => it should fail
- read a proof => should work
- attempt to close the db => it should fail
- close the block context => should work
- attempt to close the db => should work

File: go/carmen/database_test.go [hunk 18]
```diff
@@ -845,6 +845,374 @@ func TestDatabase_CloseDB_Unfinished_Queries(t *testing.T) {
...
+	block, err := db.GetHistoricContext(1)
+	if err != nil {
+		t.Fatalf("cannot get block: %v", err)
+	}
+
+	if err := block.Close(); err != nil {
+		t.Fatalf("cannot close block: %v", err)
+	}
+
+	if err := db.Close(); err != nil {
```
Comment (HerbertJordan) [inline go/carmen/database_test.go:1189]: I think the `db.Close()` should be moved after the attempt to fetch a proof since the closing of the block context needs to prevent further proof reading, not the closing of the DB.

**Review by HerbertJordan — APPROVED** [review]

File: go/database/mpt/archive_trie.go [hunk 19]
```diff
@@ -241,6 +242,13 @@ func (a *ArchiveTrie) GetHash(block uint64) (hash common.Hash, err error) {
 	return res, nil
 }
 
+func (a *ArchiveTrie) CreateWitnessProof(block uint64, address common.Address, keys ...common.Key) (witness.Proof, error) {
+	a.rootsMutex.Lock()
```
[the remark this answers is not in the document]
Comment (kjezek) [reply go/database/mpt/archive_trie.go:246]: added

File: go/carmen/database_test.go (hunk 17 above)
Comment (kjezek) [reply go/carmen/database_test.go:1142]: ok, right, changed

File: go/carmen/database_test.go (hunk 18 above)
Comment (kjezek) [reply go/carmen/database_test.go:1189]: right, moved

File: go/backend/archive/archive_test.go [hunk 20]
```diff
@@ -261,6 +262,72 @@ func TestAccountDeleteCreate(t *testing.T) {
...
+			defer func() {
+				if err := a.Close(); err != nil {
+					t.Fatalf("failed to close archive; %s", err)
+				}
+			}()
+
+			if err := a.Add(1, common.Update{
+				CreatedAccounts: []common.Address{{1}},
+				Balances: []common.BalanceUpdate{
+					{Account: common.Address{1}, Balance: common.Balance{0x12}},
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/backend/archive/archive_test.go:278]: Witness proofs have priority I'd say.

**Review by HerbertJordan — APPROVED** [review]

File: go/backend/archive/archive.go [hunk 21]
```diff
@@ -14,8 +14,12 @@ package archive
 
 import (
 	"github.com/Fantom-foundation/Carmen/go/common"
+	"github.com/Fantom-foundation/Carmen/go/common/witness"
 )
 
+// ErrWitnessProfNotSupported is returned when the archive does not support witness proofs.
+const ErrWitnessProfNotSupported = common.ConstError("witness proof not supported")
```
Comment (HerbertJordan) [inline go/backend/archive/archive.go:21]: I guess it should be `Proof` nit `Prof` ... although a `Prof` can also be a Witness :)
```suggestion
const ErrWitnessProofNotSupported = common.ConstError("witness proof not supported")
```

**Review by HerbertJordan — DISMISSED** [review]

**Review by HerbertJordan — APPROVED** [review]
